src/rules/oneLineRule.ts

```diff
@@ -27,80 +25,212 @@ export class Rule extends Lint.Rules.AbstractRule {
     public static ELSE_FAILURE_STRING = "misplaced 'else'";
     public static WHITESPACE_FAILURE_STRING = "missing whitespace";
 
-    public apply(syntaxTree: TypeScript.SyntaxTree): Lint.RuleFailure[] {
-        var braceWalker = new BraceWalker(syntaxTree, this.getOptions());
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        var braceWalker = new BraceWalker(sourceFile, this.getOptions());
         return this.applyWithWalker(braceWalker);
     }
 }
 
-class BraceWalker extends Lint.StateAwareRuleWalker {
-    public visitToken(token: TypeScript.ISyntaxToken): void {
-        var failure: Lint.RuleFailure;
-        var kind = token.kind();
-        var lastState = this.getLastState();
-
-        if (kind === TypeScript.SyntaxKind.OpenBraceToken && lastState !== undefined) {
-            var lastKind = lastState.token.kind();
-            if (lastKind === TypeScript.SyntaxKind.CloseParenToken ||
-                lastKind === TypeScript.SyntaxKind.DoKeyword ||
-                lastKind === TypeScript.SyntaxKind.ElseKeyword ||
-                lastKind === TypeScript.SyntaxKind.IdentifierName ||
-                lastKind === TypeScript.SyntaxKind.StringLiteral ||
-                lastKind === TypeScript.SyntaxKind.TryKeyword ||
-                lastKind === TypeScript.SyntaxKind.EqualsToken) {
-
-                var lastLine = this.getLine(lastState.position);
-                var currentLine = this.getLine(this.getPosition());
-                var position = this.getPosition() + TypeScript.leadingTriviaWidth(token);
-
-                if (this.hasOption(OPTION_BRACE) && currentLine !== lastLine) {
-                    failure = this.createFailure(position, TypeScript.width(token), Rule.BRACE_FAILURE_STRING);
-                } else if (this.hasOption(OPTION_WHITESPACE) && !this.hasTrailingWhiteSpace(lastState.token)) {
-                    failure = this.createFailure(position, TypeScript.width(token), Rule.WHITESPACE_FAILURE_STRING);
+class BraceWalker extends Lint.RuleWalker {
+    public visitIfStatement(node: ts.IfStatement) {
+        var sourceFile = node.getSourceFile();
+        var thenStatement = node.thenStatement;
+        if (thenStatement.kind === ts.SyntaxKind.Block) {
+            var expressionCloseParen = node.getChildAt(3);
+            var thenOpeningBrace = thenStatement.getChildAt(0);
+            this.handleOpeningBrace(expressionCloseParen, thenOpeningBrace);
+        }
+
+        var elseStatement = node.elseStatement;
+        if (node.elseStatement != null) {
```

Comment:
also do an early return here if elseStatement is null, it's cleaner to read code below because it won't be nested